### PR TITLE
Feat trait to make clients customizable

### LINKED ISSUE #452: Let users customize their web client
Following: https://github.com/meilisearch/meilisearch-rust/pull/426#issuecomment-1466375016

It would be nice to make the web client customizable, i.e:
- Add a feature flag to get rid of `ishac` (by default, we should continue to use it)
- Add a `new_with_client` function to create a [`Client`](https://github.com/meilisearch/meilisearch-rust/blob/9b0eafe32101ea9e8670de4c1407bfe6e411b0e8/src/client.rs#L17-L21) with anything that implements [`request`](https://github.com/meilisearch/meilisearch-rust/blob/9b0eafe32101ea9e8670de4c1407bfe6e411b0e8/src/request.rs#L118-L210)

### PR DESCRIPTION
**Pull request by shimatar0**
# Pull Request

## Related issue
Fixes #452 

## What does this PR do?
- This is a WIP state and I'll add it next time.
- There are still many problems to be solved. I will comment on the code.

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?
- [x] Fix the settings macro
- [x] Fix compiling without default features

Thank you so much for contributing to Meilisearch!

### CODE REVIEW

File: src/tasks.rs
Comment (shimatar0) [inline src/tasks.rs:189]: compiler error
```
error[E0392]: parameter `Http` is never used
   --> src/tasks.rs:189:15
    |
189 | pub enum Task<Http: HttpClient> {
    |               ^^^^ unused parameter
    |
    = help: consider removing `Http`, referring to it in a field, or using a marker such as `PhantomData`
```

However, Task feels that this parameter is necessary to call Client
https://github.com/meilisearch/meilisearch-rust/blob/8a4f62d49a2433a1db6d1e06426fb27731403ce5/src/tasks.rs#L263

**Comment by shimatar0** [comment]
Sorry. It's not in working condition at all yet. It may be a more difficult task than think. 
It is very likely that this PR will only share the situation and not finalize the merge.

File: src/tasks.rs
Comment (shimatar0) [reply src/tasks.rs:189]: Similar compile errors have been reported by others...

File: src/client.rs
Comment (shimatar0) [inline src/client.rs:49]: This could be an elementary problem...
It seems to me that you are fulfilling.
```
mismatched types
expected type parameter `Http`
           found struct `request::IsahcClinet`
```

**Comment by shimatar0** [comment]
This sounds like a lot to think about!
Return value needed to be changed to meet async trait.
So it looks like we need to match the `parse_response`.
```
error[E0277]: the trait bound `dyn futures::Future<Output = Result<Box<dyn std::marker::Send>, errors::Error>> + std::marker::Send: Deserialize<'_>` is not satisfied
   --> src/request.rs:123:9
    |
123 |         parse_response(status, expected_status_code, body)
    |         ^^^^^^^^^^^^^^ the trait `Deserialize<'_>` is not implemented for `dyn futures::Future<Output = Result<Box<dyn std::marker::Send>, errors::Error>> + std::marker::Send`
```

File: src/lib.rs
Comment (shimatar0) [inline src/lib.rs:238]: I've had to expose the request mod because in order to use HttpClient

File: src/documents.rs
Comment (shimatar0) [inline src/documents.rs:57]: I needed it to build with wasm32.
But this might not be a very good way to write it.

File: src/dumps.rs
Comment (shimatar0) [inline src/dumps.rs:74]: Clone is now required for the call.
And other places.

File: examples/cli-app-with-reqwest/Cargo.toml
Comment (irevoire) [inline examples/cli-app-with-reqwest/Cargo.toml:2]:
```suggestion
name = "cli-app-with-reqwest"
```

File: examples/cli-app-with-reqwest/src/main.rs
Comment (irevoire) [inline examples/cli-app-with-reqwest/src/main.rs:132]:
```suggestion
        unimplemented!("stream_request is not implemented for ReqwestClient")
```

File: src/request.rs
Comment (irevoire) [inline src/request.rs:42]:
```suggestion
    async fn request<Query, Body, Output>(
        self,
        url: &str,
        apikey: Option<&str>,
        method: Method<Query, Body>,
        expected_status_code: u16,
    ) -> Result<Output, Error>
    where
        Query: Serialize + Send + Sync,
        Body: Serialize + Send + Sync,
        Output: DeserializeOwned + 'static + Send;
```

Huum, I’m struggling to see a difference between these two 🤔 
Shouldn’t we be able to simply define it once without the wasm/no-wasm attribute?

File: src/request.rs
Comment (irevoire) [inline src/request.rs:60]:
```suggestion
#[derive(Debug, Clone, Copy, Default, Serialize)]
```

File: src/request.rs
Comment (irevoire) [inline src/request.rs:73]:
```suggestion
```

Since we can derive `Default` we don’t need this anymore

File: src/request.rs
Comment (irevoire) [inline src/request.rs:84]: Here it would be better to split the two implementations in different block I think;

```rust
// First, the non-wasm impl block where the whole block is gated behind the cfg thingy

#[cfg(not(target_arch = "wasm32"))]
#[async_trait(?Send)]
impl HttpClient for IsahcClient {
async fn request<Query, Body, Output>(
        self,
        url: &str,
       ...) -> Result<Output, Error> { ... }

    async fn stream_request<
        'a,
        Query: Serialize + Send + Sync,
        Body: futures_io::AsyncRead + Send + Sync + 'static,
        Output: DeserializeOwned + 'static,
    >(
        self,
        url: &str,
        ...) -> Result<Output, Error> { ... }
}

// And then the wasm block that doesn’t contains the stream_request method at all

#[cfg(target_arch = "wasm32")]
#[async_trait(?Send)]
impl HttpClient for IsahcClient {
async fn request<Query, Body, Output>(
        self,
        url: &str,
       ...) -> Result<Output, Error> { ... }
}
```

What do you think?

**Review by irevoire — CHANGES_REQUESTED** [review]
You have some warning on your new example

File: examples/cli-app-with-reqwest/Cargo.toml
Comment (shimatar0) [reply examples/cli-app-with-reqwest/Cargo.toml]: Fixed
https://github.com/meilisearch/meilisearch-rust/pull/459/commits/1f722a791adecf46def0234913ae18da85479747

File: examples/cli-app-with-reqwest/src/main.rs
Comment (shimatar0) [reply examples/cli-app-with-reqwest/src/main.rs]: Fixed
https://github.com/meilisearch/meilisearch-rust/pull/459/commits/1f722a791adecf46def0234913ae18da85479747

File: src/request.rs
Comment (shimatar0) [reply src/request.rs]: You are right!
I didn't bother to separate them 😂
fixed
https://github.com/meilisearch/meilisearch-rust/pull/459/commits/2a1cc7a8e6b088c15c7c34ac3fe98246c1b96947

File: src/request.rs
Comment (shimatar0) [reply src/request.rs]: fixed
https://github.com/meilisearch/meilisearch-rust/pull/459/commits/e7e2c355b1e410a02cb81d06cb4d9fa0f6ec9b9c

File: src/request.rs
Comment (shimatar0) [reply src/request.rs]: fixed
https://github.com/meilisearch/meilisearch-rust/pull/459/commits/e7e2c355b1e410a02cb81d06cb4d9fa0f6ec9b9c

File: src/request.rs
Comment (shimatar0) [reply src/request.rs]: It certainly doesn't feel right to have methods with the same name in the same block.
I think it is appropriate to separate them by `flag` and keep the necessary ones, so I've modified it!🙏

https://github.com/meilisearch/meilisearch-rust/pull/459/commits/2a1cc7a8e6b088c15c7c34ac3fe98246c1b96947

File: examples/cli-app-with-reqwest/Cargo.toml
Comment (bidoubiwa) [inline examples/cli-app-with-reqwest/Cargo.toml:17]:
```suggestion
tokio = { version = "1.27.0",  features = ["full"] }

```

File: examples/cli-app-with-reqwest/Cargo.toml
Comment (bidoubiwa) [inline examples/cli-app-with-reqwest/Cargo.toml:10]: Here you could the add `no-default-feature` option

File: examples/cli-app-with-reqwest/Cargo.toml
Comment (bidoubiwa) [inline examples/cli-app-with-reqwest/Cargo.toml:12]:
```suggestion
serde = { version="1.0", features = ["derive"] }
```

File: examples/cli-app-with-reqwest/Cargo.toml
Comment (bidoubiwa) [inline examples/cli-app-with-reqwest/Cargo.toml:17]:
```suggestion
tokio = {version = "1.27.0", features = ["full"]}
```

File: examples/settings.rs
Comment (bidoubiwa) [inline examples/settings.rs:12]: This is a huge breaking and impacts the user experience considerably :( 

Can we have the `isahcClient` by default which you only would have to declare if you are in `no default feature` mode?

File: src/client.rs
Comment (bidoubiwa) [inline src/client.rs:147]: Is it possible to hide the http client here so that the user does not have to provide it if they use ishac?

File: src/request.rs
Comment (bidoubiwa) [inline src/request.rs:258]: This can be removed, isahc does not work in wasm

File: src/request.rs
Comment (bidoubiwa) [inline src/request.rs:261]: not sure this work

File: examples/cli-app-with-reqwest/Cargo.toml
Comment (shimatar0) [reply examples/cli-app-with-reqwest/Cargo.toml]: fixed
https://github.com/meilisearch/meilisearch-rust/pull/459/commits/f7986c7031912da09b09fdbc07366f6b668dcfa1

File: examples/cli-app-with-reqwest/Cargo.toml
Comment (shimatar0) [reply examples/cli-app-with-reqwest/Cargo.toml]: fixed
https://github.com/meilisearch/meilisearch-rust/pull/459/commits/daa85e8835be2b272bb6f2cd7d66e733648e7c8b

File: examples/cli-app-with-reqwest/Cargo.toml
Comment (shimatar0) [reply examples/cli-app-with-reqwest/Cargo.toml]: changed
https://github.com/meilisearch/meilisearch-rust/pull/459/commits/35221ea4d4d3850a5d59a169dd2757441439c803

File: examples/settings.rs
Comment (shimatar0) [reply examples/settings.rs]: You are right. Thanks for checking and taking the time to help me out 😊
This was a destructive change for users who had specified the type :(

I modified the SDK so that users who use it by default can use it without worrying about traits.(but a little forced😢)
However, I'm not sure if this is a good thing to do, as it also requires changes to the export and to `lib.rs`. 

All in all, this is a big change 
so I hope the maintainer team will consider the decision and any better response 🙏

https://github.com/meilisearch/meilisearch-rust/pull/459/commits/6d96c9f956b6ef02e6a92b41b017d5a06eb840e0

Comment (bidoubiwa) [reply examples/settings.rs]: What do you think @irevoire ?

Comment (irevoire) [reply examples/settings.rs]: Yeah, I don't see any other way to do it tbh.
What would you think of having a `prelude` like they have in some libs @bidoubiwa?
For example, `rayon` uses one (look at their examples. They all starts with `use rayon::prelude::*;` instead of picking each thing you need separately: https://docs.rs/rayon/latest/rayon/iter/trait.ParallelIterator.html#method.for_each_with
serenity also exposes a prelude: https://docs.rs/serenity/latest/serenity/client/struct.Client.html

I feel like that's the easiest way for us to define all the types we need to overwrite in a prelude file gated by the feature being used.

That being said, it's a « big » breaking change, it'll probably break a lot of code, but the fix for the end users will be way easier => get rids of all your imports and just write `use meilisearch_sdk::prelude::*;`
The cool thing for us is that we can easily change the type alias in the prelude file without being breaking in the future + create a prelude for the wasm + a prelude for isahc

File: src/request.rs
Comment (shimatar0) [reply src/request.rs]: Fixed to use `WebSysClient` for wasm.

https://github.com/meilisearch/meilisearch-rust/commit/71f2abebb91da8e9b3afe29745730804f7f869ec

File: src/request.rs
Comment (shimatar0) [reply src/request.rs]: Removed this

https://github.com/meilisearch/meilisearch-rust/commit/71f2abebb91da8e9b3afe29745730804f7f869ec

File: examples/settings.rs
Comment (shimatar0) [reply examples/settings.rs]: I tried to modify the form to use `prelude`!
Please let me know if I'm wrong🙏

https://github.com/meilisearch/meilisearch-rust/pull/459/commits/bd04a9bd5be75cdd03b43b0dc0ed92b4c6bbe199
https://github.com/meilisearch/meilisearch-rust/pull/459/commits/5b326278139e521261c659a1a0e83a84890cb1ae

File: src/client.rs
Comment (shimatar0) [reply src/client.rs:176]: I have thought about it a bit, but it seems difficult due to the presence of the Client in the `MultiSearchQuery` structure🤔
If the use case of `MultiSearchQuery` is only to be called as a method from the `Client` structure, I don't think there is that much of a problem. 
However, if you are going to implement `MultiSearchQuery` directly, I think it would be better to modify it.

File: Cargo.toml
Comment (shimatar0) [reply Cargo.toml:41]: Added description to readme.
I'm a little unsure whether to write a code example or not in here, 
because writing something that moves is going to be huge!🤔

https://github.com/meilisearch/meilisearch-rust/pull/459/commits/5993c50552fa5f7a4ca63edf5dc3657b6800ce9b

**Comment by novacrazy** [comment]
What is the status of this? 

I'm looking into using Meilisearch, but having to bundle an entire copy of curl into my binary just because of this issue is giving me hesitation. I use hyper/reqwest extensively within my server already.

**Comment by irevoire** [comment]
The pr is mainly finished. It should go over a last round of review, but with the MeiliSearch release coming out this Monday, we didn't find the time to merge it.
I hope we’ll be able to merge it next week

**Review by bidoubiwa — CHANGES_REQUESTED** [review]
Hey! @shimatar0. Sorry for the delay.

Thanks a lot again for the great work. These are the changes we would like to see in your PR:

- Add all the public types in the preludes (DocumentsQuery, Settings, etc)
- Update the readme using the `prelude` import.
- Add a section in the readme mentioning the custom HTTP feature, with a link pointing to your cli app with reqwest.
- Update the code-samples `.code-samples.meilisearch.yml`

**Comment by gibbz00** [comment]
Is there anything I can do to get this PR past the finishing line? I'm interested in adding a `reqwest` client backend but that contribution would be blocked by this.

File: examples/cli-app-with-reqwest/src/main.rs
Comment (irevoire) [inline examples/cli-app-with-reqwest/src/main.rs:108]:
```suggestion
```

File: src/client.rs
Comment (irevoire) [reply src/client.rs:178]: Done

File: src/client.rs
Comment (irevoire) [inline src/client.rs:220]:
```suggestion
    /// #
```

**Review by irevoire — CHANGES_REQUESTED** [review]

**Pull request merged**
